File: test/javascript/erlang/uri_string_test.mjs
"use strict";

import {
  assert,
  assertBoxedError,
  defineGlobalErlangAndElixirModules,
} from "../support/helpers.mjs";

import Erlang_Uri_String from "../../../assets/js/erlang/uri_string.mjs";
import Interpreter from "../../../assets/js/interpreter.mjs";
import Type from "../../../assets/js/type.mjs";

defineGlobalErlangAndElixirModules();

// IMPORTANT!
// Each JavaScript test has a related Elixir consistency test in test/elixir/hologram/ex_js_consistency/erlang/uri_string_test.exs
// Always update both together.

describe("Erlang_Uri_String", () => {
  describe("parse/1", () => {
    const parse = Erlang_Uri_String["parse/1"];

    it("full URI with all components (binary)", () => {
      const uri = Type.bitstring(
        "foo://user@example.com:8042/over/there?name=ferret#nose",
      );
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("fragment"), Type.bitstring("nose")],
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/over/there")],
        [Type.atom("port"), Type.integer(8042)],
        [Type.atom("query"), Type.bitstring("name=ferret")],
        [Type.atom("scheme"), Type.bitstring("foo")],
        [Type.atom("userinfo"), Type.bitstring("user")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("full URI with all components (list)", () => {
      const uri = Type.charlist(
        "foo://user@example.com:8042/over/there?name=ferret#nose",
      );
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("fragment"), Type.charlist("nose")],
        [Type.atom("host"), Type.charlist("example.com")],
        [Type.atom("path"), Type.charlist("/over/there")],
        [Type.atom("port"), Type.integer(8042)],
        [Type.atom("query"), Type.charlist("name=ferret")],
        [Type.atom("scheme"), Type.charlist("foo")],
        [Type.atom("userinfo"), Type.charlist("user")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI without userinfo", () => {
      const uri = Type.bitstring(
        "foo://example.com:8042/over/there?name=ferret#nose",
      );
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("fragment"), Type.bitstring("nose")],
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/over/there")],
        [Type.atom("port"), Type.integer(8042)],
        [Type.atom("query"), Type.bitstring("name=ferret")],
        [Type.atom("scheme"), Type.bitstring("foo")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI without port", () => {
      const uri = Type.bitstring(
        "foo://user@example.com/over/there?name=ferret#nose",
      );
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("fragment"), Type.bitstring("nose")],
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/over/there")],
        [Type.atom("query"), Type.bitstring("name=ferret")],
        [Type.atom("scheme"), Type.bitstring("foo")],
        [Type.atom("userinfo"), Type.bitstring("user")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI without query", () => {
      const uri = Type.bitstring("foo://user@example.com:8042/over/there#nose");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("fragment"), Type.bitstring("nose")],
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/over/there")],
        [Type.atom("port"), Type.integer(8042)],
        [Type.atom("scheme"), Type.bitstring("foo")],
        [Type.atom("userinfo"), Type.bitstring("user")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI without fragment", () => {
      const uri = Type.bitstring(
        "foo://user@example.com:8042/over/there?name=ferret",
      );
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/over/there")],
        [Type.atom("port"), Type.integer(8042)],
        [Type.atom("query"), Type.bitstring("name=ferret")],
        [Type.atom("scheme"), Type.bitstring("foo")],
        [Type.atom("userinfo"), Type.bitstring("user")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("simple HTTP URI", () => {
      const uri = Type.bitstring("http://www.example.com");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("www.example.com")],
        [Type.atom("path"), Type.bitstring("")],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("HTTP URI with path", () => {
      const uri = Type.bitstring("http://www.example.com/path/to/resource");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("www.example.com")],
        [Type.atom("path"), Type.bitstring("/path/to/resource")],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("HTTPS URI with port and query", () => {
      const uri = Type.bitstring("https://example.com:443/search?q=test");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/search")],
        [Type.atom("port"), Type.integer(443)],
        [Type.atom("query"), Type.bitstring("q=test")],
        [Type.atom("scheme"), Type.bitstring("https")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI with IPv6 address", () => {
      const uri = Type.bitstring("http://[2001:db8::1]:8080/path");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("2001:db8::1")],
        [Type.atom("path"), Type.bitstring("/path")],
        [Type.atom("port"), Type.integer(8080)],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI with IPv6 address without port", () => {
      const uri = Type.bitstring("http://[2001:db8::1]/path");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("2001:db8::1")],
        [Type.atom("path"), Type.bitstring("/path")],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI with empty port", () => {
      const uri = Type.bitstring("http://example.com:/path");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/path")],
        [Type.atom("port"), Type.atom("undefined")],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI with port 0", () => {
      const uri = Type.bitstring("http://example.com:0/path");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/path")],
        [Type.atom("port"), Type.integer(0)],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("relative URI (path only)", () => {
      const uri = Type.bitstring("/path/to/resource");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("path"), Type.bitstring("/path/to/resource")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("relative URI with query", () => {
      const uri = Type.bitstring("/path?query=value");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("path"), Type.bitstring("/path")],
        [Type.atom("query"), Type.bitstring("query=value")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("relative URI with fragment", () => {
      const uri = Type.bitstring("/path#section");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("path"), Type.bitstring("/path")],
        [Type.atom("fragment"), Type.bitstring("section")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("fragment only", () => {
      const uri = Type.bitstring("#fragment");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("path"), Type.bitstring("")],
        [Type.atom("fragment"), Type.bitstring("fragment")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("query only", () => {
      const uri = Type.bitstring("?query=value");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("path"), Type.bitstring("")],
        [Type.atom("query"), Type.bitstring("query=value")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("empty URI string", () => {
      const uri = Type.bitstring("");
      const result = parse(uri);

      const expected = Type.map([[Type.atom("path"), Type.bitstring("")]]);

      assert.deepStrictEqual(result, expected);
    });

    it("mailto URI", () => {
      const uri = Type.bitstring("mailto:john@example.com");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("path"), Type.bitstring("john@example.com")],
        [Type.atom("scheme"), Type.bitstring("mailto")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("file URI", () => {
      const uri = Type.bitstring("file:///path/to/file.txt");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("")],
        [Type.atom("path"), Type.bitstring("/path/to/file.txt")],
        [Type.atom("scheme"), Type.bitstring("file")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI with percent-encoded characters", () => {
      const uri = Type.bitstring("http://example.com/path%20with%20spaces");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/path%20with%20spaces")],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI with special characters in query", () => {
      const uri = Type.bitstring("http://example.com?key1=value1&key2=value2");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("")],
        [Type.atom("query"), Type.bitstring("key1=value1&key2=value2")],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("URI with userinfo containing password", () => {
      const uri = Type.bitstring("http://user:password@example.com/path");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("example.com")],
        [Type.atom("path"), Type.bitstring("/path")],
        [Type.atom("scheme"), Type.bitstring("http")],
        [Type.atom("userinfo"), Type.bitstring("user:password")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("invalid UTF-8 binary raises FunctionClauseError", () => {
      const invalid = Type.bitstring([1, 1, 1, 1, 1, 1, 1, 1]);

      assertBoxedError(
        () => parse(invalid),
        "FunctionClauseError",
        Interpreter.buildFunctionClauseErrorMsg(":uri_string.parse/1", [
          invalid,
        ]),
      );
    });

    it("charlist with non-integer raises ArgumentError", () => {
      const invalid = Type.list([Type.atom("a")]);

      assertBoxedError(
        () => parse(invalid),
        "ArgumentError",
        /not valid character data/, // matches OTP-style message
      );
    });

    it("charlist with out-of-range integer raises FunctionClauseError", () => {
      const invalid = Type.list([Type.integer(1_200_000)]);

      assertBoxedError(
        () => parse(invalid),
        "FunctionClauseError",
        Interpreter.buildFunctionClauseErrorMsg(":uri_string.parse/1", [
          invalid,
        ]),
      );
    });

    it("URI with IPv6 host and port", () => {
      const uri = Type.bitstring("http://[::1]:8080/");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("::1")],
        [Type.atom("path"), Type.bitstring("/")],
        [Type.atom("port"), Type.integer(8080)],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("bare authority with empty host and port", () => {
      const uri = Type.bitstring("//:80");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("")],
        [Type.atom("path"), Type.bitstring("")],
        [Type.atom("port"), Type.integer(80)],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("missing closing bracket returns invalid_uri", () => {
      const uri = Type.bitstring("http://[::1");
      const result = parse(uri);

      const expected = Type.tuple([
        Type.atom("error"),
        Type.atom("invalid_uri"),
        Type.charlist(":"),
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("non-numeric port returns invalid_uri", () => {
      const uri = Type.bitstring("http://example.com:abc");
      const result = parse(uri);

      const expected = Type.tuple([
        Type.atom("error"),
        Type.atom("invalid_uri"),
        Type.charlist(":"),
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("second fragment separator returns invalid_uri", () => {
      const uri = Type.bitstring("http://h/p#f#x");
      const result = parse(uri);

      const expected = Type.tuple([
        Type.atom("error"),
        Type.atom("invalid_uri"),
        Type.charlist(":"),
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("double query keeps full query text", () => {
      const uri = Type.bitstring("http://h/p?x?y");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("h")],
        [Type.atom("path"), Type.bitstring("/p")],
        [Type.atom("query"), Type.bitstring("x?y")],
        [Type.atom("scheme"), Type.bitstring("http")],
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("raises FunctionClauseError for non-binary/non-list input", () => {
      const invalidInput = Type.integer(123);

      assertBoxedError(
        () => parse(invalidInput),
        "FunctionClauseError",
        Interpreter.buildFunctionClauseErrorMsg(":uri_string.parse/1", [
          invalidInput,
        ]),
      );
    });

    it("multiple @ symbols in authority (bare //) returns invalid_uri", () => {
      const uri = Type.bitstring("//a@b@c/path");
      const result = parse(uri);

      const expected = Type.tuple([
        Type.atom("error"),
        Type.atom("invalid_uri"),
        Type.charlist("@"),
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("multiple @ symbols in authority (with scheme) returns invalid_uri", () => {
      const uri = Type.bitstring("http://a@b@c/path");
      const result = parse(uri);

      const expected = Type.tuple([
        Type.atom("error"),
        Type.atom("invalid_uri"),
        Type.charlist(":"),
      ]);

      assert.deepStrictEqual(result, expected);
    });

    it("multiple @ symbols in authority with port returns invalid_uri", () => {
      const uri = Type.bitstring("http://user@host@extra:8080/path");
      const result = parse(uri);

      const expected = Type.tuple([
        Type.atom("error"),
        Type.atom("invalid_uri"),
        Type.charlist(":"),
      ]);

      assert.deepStrictEqual(result, expected);
    });

    // Duplicate of the bare // case above removed for clarity.

    it("single @ in userinfo, multiple in path is valid", () => {
      const uri = Type.bitstring("http://user@host/path@with@at");
      const result = parse(uri);

      const expected = Type.map([
        [Type.atom("host"), Type.bitstring("host")],
        [Type.atom("path"), Type.bitstring("/path@with@at")],
        [Type.atom("scheme"), Type.bitstring("http")],
        [Type.atom("userinfo"), Type.bitstring("user")],
      ]);

      assert.deepStrictEqual(result, expected);
    });
  });
});

File: test/elixir/hologram/ex_js_consistency/erlang/uri_string_test.exs
defmodule Hologram.ExJsConsistency.Erlang.UriStringTest do
  @moduledoc """
  IMPORTANT!
  Each Elixir consistency test has a related JavaScript test in test/javascript/erlang/uri_string_test.mjs
  Always update both together.
  """

  use Hologram.Test.BasicCase, async: true

  @moduletag :consistency

  describe "parse/1" do
    test "full URI with all components (binary)" do
      assert :uri_string.parse("foo://user@example.com:8042/over/there?name=ferret#nose") ==
               %{
                 fragment: "nose",
                 host: "example.com",
                 path: "/over/there",
                 port: 8042,
                 query: "name=ferret",
                 scheme: "foo",
                 userinfo: "user"
               }
    end

    test "full URI with all components (list)" do
      assert :uri_string.parse(~c"foo://user@example.com:8042/over/there?name=ferret#nose") ==
               %{
                 fragment: ~c"nose",
                 host: ~c"example.com",
                 path: ~c"/over/there",
                 port: 8042,
                 query: ~c"name=ferret",
                 scheme: ~c"foo",
                 userinfo: ~c"user"
               }
    end

    test "URI without userinfo" do
      assert :uri_string.parse("foo://example.com:8042/over/there?name=ferret#nose") ==
               %{
                 fragment: "nose",
                 host: "example.com",
                 path: "/over/there",
                 port: 8042,
                 query: "name=ferret",
                 scheme: "foo"
               }
    end

    test "URI without port" do
      assert :uri_string.parse("foo://user@example.com/over/there?name=ferret#nose") ==
               %{
                 fragment: "nose",
                 host: "example.com",
                 path: "/over/there",
                 query: "name=ferret",
                 scheme: "foo",
                 userinfo: "user"
               }
    end

    test "URI without query" do
      assert :uri_string.parse("foo://user@example.com:8042/over/there#nose") ==
               %{
                 fragment: "nose",
                 host: "example.com",
                 path: "/over/there",
                 port: 8042,
                 scheme: "foo",
                 userinfo: "user"
               }
    end

    test "URI without fragment" do
      assert :uri_string.parse("foo://user@example.com:8042/over/there?name=ferret") ==
               %{
                 host: "example.com",
                 path: "/over/there",
                 port: 8042,
                 query: "name=ferret",
                 scheme: "foo",
                 userinfo: "user"
               }
    end

    test "simple HTTP URI" do
      assert :uri_string.parse("http://www.example.com") ==
               %{
                 host: "www.example.com",
                 path: "",
                 scheme: "http"
               }
    end

    test "HTTP URI with path" do
      assert :uri_string.parse("http://www.example.com/path/to/resource") ==
               %{
                 host: "www.example.com",
                 path: "/path/to/resource",
                 scheme: "http"
               }
    end

    test "HTTPS URI with port and query" do
      assert :uri_string.parse("https://example.com:443/search?q=test") ==
               %{
                 host: "example.com",
                 path: "/search",
                 port: 443,
                 query: "q=test",
                 scheme: "https"
               }
    end

    test "URI with IPv6 address" do
      assert :uri_string.parse("http://[2001:db8::1]:8080/path") ==
               %{
                 host: "2001:db8::1",
                 path: "/path",
                 port: 8080,
                 scheme: "http"
               }
    end

    test "URI with IPv6 address without port" do
      assert :uri_string.parse("http://[2001:db8::1]/path") ==
               %{
                 host: "2001:db8::1",
                 path: "/path",
                 scheme: "http"
               }
    end

    test "URI with empty port" do
      assert :uri_string.parse("http://example.com:/path") ==
               %{
                 host: "example.com",
                 path: "/path",
                 port: :undefined,
                 scheme: "http"
               }
    end

    test "relative URI (path only)" do
      assert :uri_string.parse("/path/to/resource") ==
               %{
                 path: "/path/to/resource"
               }
    end

    test "relative URI with query" do
      assert :uri_string.parse("/path?query=value") ==
               %{
                 path: "/path",
                 query: "query=value"
               }
    end

    test "relative URI with fragment" do
      assert :uri_string.parse("/path#section") ==
               %{
                 path: "/path",
                 fragment: "section"
               }
    end

    test "fragment only" do
      assert :uri_string.parse("#fragment") ==
               %{
                 path: "",
                 fragment: "fragment"
               }
    end

    test "query only" do
      assert :uri_string.parse("?query=value") ==
               %{
                 path: "",
                 query: "query=value"
               }
    end

    test "empty URI string" do
      assert :uri_string.parse("") ==
               %{
                 path: ""
               }
    end

    test "mailto URI" do
      assert :uri_string.parse("mailto:john@example.com") ==
               %{
                 path: "john@example.com",
                 scheme: "mailto"
               }
    end

    test "file URI" do
      assert :uri_string.parse("file:///path/to/file.txt") ==
               %{
                 host: "",
                 path: "/path/to/file.txt",
                 scheme: "file"
               }
    end

    test "URI with percent-encoded characters" do
      assert :uri_string.parse("http://example.com/path%20with%20spaces") ==
               %{
                 host: "example.com",
                 path: "/path%20with%20spaces",
                 scheme: "http"
               }
    end

    test "URI with special characters in query" do
      assert :uri_string.parse("http://example.com?key1=value1&key2=value2") ==
               %{
                 host: "example.com",
                 path: "",
                 query: "key1=value1&key2=value2",
                 scheme: "http"
               }
    end

    test "URI with userinfo containing password" do
      assert :uri_string.parse("http://user:password@example.com/path") ==
               %{
                 host: "example.com",
                 path: "/path",
                 scheme: "http",
                 userinfo: "user:password"
               }
    end

    test "invalid UTF-8 binary raises FunctionClauseError" do
      assert_error FunctionClauseError,
                   ~r/parse_scheme_start/,
                   fn -> :uri_string.parse(<<255>>) end
    end

    test "charlist with non-integer raises ArgumentError" do
      assert_error ArgumentError,
                   ~r/not valid character data/,
                   fn -> :uri_string.parse([:a]) end
    end

    test "charlist with out-of-range integer raises FunctionClauseError" do
      assert_error FunctionClauseError,
                   ~r/parse_scheme_start/,
                   fn -> :uri_string.parse([1_200_000]) end
    end

    test "URI with IPv6 host and port" do
      assert :uri_string.parse("http://[::1]:8080/") ==
               %{
                 host: "::1",
                 path: "/",
                 port: 8080,
                 scheme: "http"
               }
    end

    test "bare authority with empty host and port" do
      assert :uri_string.parse("//:80") ==
               %{
                 host: "",
                 path: "",
                 port: 80
               }
    end

    test "missing closing bracket returns invalid_uri" do
      assert :uri_string.parse("http://[::1") == {:error, :invalid_uri, ~c":"}
    end

    test "non-numeric port returns invalid_uri" do
      assert :uri_string.parse("http://example.com:abc") == {:error, :invalid_uri, ~c":"}
    end

    test "second fragment separator returns invalid_uri" do
      assert :uri_string.parse("http://h/p#f#x") == {:error, :invalid_uri, ~c":"}
    end

    test "double query keeps full query text" do
      assert :uri_string.parse("http://h/p?x?y") ==
               %{
                 host: "h",
                 path: "/p",
                 query: "x?y",
                 scheme: "http"
               }
    end

    test "raises FunctionClauseError for non-binary/non-list input" do
      expected_msg = build_function_clause_error_msg(":uri_string.parse/1", [123])

      assert_error FunctionClauseError,
                   expected_msg,
                   fn -> :uri_string.parse(123) end
    end

    test "multiple @ symbols in authority (bare //) returns invalid_uri" do
      assert :uri_string.parse("//a@b@c/path") == {:error, :invalid_uri, ~c"@"}
    end

    test "multiple @ symbols in authority (with scheme) returns invalid_uri" do
      assert :uri_string.parse("http://a@b@c/path") == {:error, :invalid_uri, ~c":"}
    end

    test "multiple @ symbols in authority with port returns invalid_uri" do
      assert :uri_string.parse("http://user@host@extra:8080/path") ==
               {:error, :invalid_uri, ~c":"}
    end

    test "single @ in userinfo, multiple in path is valid" do
      assert :uri_string.parse("http://user@host/path@with@at") ==
               %{
                 host: "host",
                 path: "/path@with@at",
                 scheme: "http",
                 userinfo: "user"
               }
    end
  end
end
